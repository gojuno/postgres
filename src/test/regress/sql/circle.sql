--
-- CIRCLE
--

CREATE TABLE CIRCLE_TBL (f1 circle);

INSERT INTO CIRCLE_TBL VALUES ('<(5,1),3>');

INSERT INTO CIRCLE_TBL VALUES ('<(1,2),100>');

INSERT INTO CIRCLE_TBL VALUES ('1,3,5');

INSERT INTO CIRCLE_TBL VALUES ('((1,2),3)');

INSERT INTO CIRCLE_TBL VALUES ('<(100,200),10>');

INSERT INTO CIRCLE_TBL VALUES ('<(100,1),115>');

-- bad values

INSERT INTO CIRCLE_TBL VALUES ('<(-100,0),-100>');

INSERT INTO CIRCLE_TBL VALUES ('1abc,3,5');

INSERT INTO CIRCLE_TBL VALUES ('(3,(1,2),3)');

SELECT * FROM CIRCLE_TBL;

SELECT '' AS six, center(f1) AS center
  FROM CIRCLE_TBL;

SELECT '' AS six, radius(f1) AS radius
  FROM CIRCLE_TBL;

SELECT '' AS six, diameter(f1) AS diameter
  FROM CIRCLE_TBL;

SELECT '' AS two, f1 FROM CIRCLE_TBL WHERE radius(f1) < 5;

SELECT '' AS four, f1 FROM CIRCLE_TBL WHERE diameter(f1) >= 10;

SELECT '' as five, c1.f1 AS one, c2.f1 AS two, (c1.f1 <-> c2.f1) AS distance
  FROM CIRCLE_TBL c1, CIRCLE_TBL c2
  WHERE (c1.f1 < c2.f1) AND ((c1.f1 <-> c2.f1) > 0)
  ORDER BY distance, area(c1.f1), area(c2.f1);

--
-- Test the SP-GiST index
--

CREATE TEMPORARY TABLE quad_circle_tbl (id int, c circle);

INSERT INTO quad_circle_tbl
	SELECT (x - 1) * 100 + y, circle(point(x * 10, y * 10), 1 + (x + y) % 10)
	FROM generate_series(1, 100) x,
		 generate_series(1, 100) y;

INSERT INTO quad_circle_tbl
	SELECT i, '<(200, 300), 5>'
	FROM generate_series(10001, 11000) AS i;

INSERT INTO quad_circle_tbl
	VALUES
		(11001, NULL),
		(11002, NULL),
		(11003, '<(0,100), infinity>'),
		(11004, '<(-infinity,0),1000>'),
		(11005, '<(infinity,-infinity),infinity>');

CREATE INDEX quad_circle_tbl_idx ON quad_circle_tbl USING spgist(c);

-- get reference results for ORDER BY distance from seq scan
SET enable_seqscan = ON;
SET enable_indexscan = OFF;
SET enable_bitmapscan = OFF;

CREATE TEMP TABLE quad_circle_tbl_ord_seq1 AS
SELECT rank() OVER (ORDER BY c <-> point '123,456') n, c <-> point '123,456' dist, id
FROM quad_circle_tbl;

CREATE TEMP TABLE quad_circle_tbl_ord_seq2 AS
SELECT rank() OVER (ORDER BY c <-> point '123,456') n, c <-> point '123,456' dist, id
FROM quad_circle_tbl WHERE c <@ circle '<(300,400),200>';

-- check results results from index scan
SET enable_seqscan = OFF;
SET enable_indexscan = OFF;
SET enable_bitmapscan = ON;

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c << circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c << circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c &< circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c &< circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c && circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c && circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c &> circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c &> circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c >> circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c >> circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c <<| circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c <<| circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c &<| circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c &<| circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c |&> circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c |&> circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c |>> circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c |>> circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c @> circle '<(300,400),1>';
SELECT count(*) FROM quad_circle_tbl WHERE c @> circle '<(300,400),1>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c <@ circle '<(300,400),200>';
SELECT count(*) FROM quad_circle_tbl WHERE c <@ circle '<(300,400),200>';

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c ~= circle '<(300,400),1>';
SELECT count(*) FROM quad_circle_tbl WHERE c ~= circle '<(300,400),1>';

RESET enable_seqscan;
RESET enable_indexscan;
RESET enable_bitmapscan;
