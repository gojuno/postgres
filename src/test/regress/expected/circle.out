--
-- CIRCLE
--
CREATE TABLE CIRCLE_TBL (f1 circle);
INSERT INTO CIRCLE_TBL VALUES ('<(5,1),3>');
INSERT INTO CIRCLE_TBL VALUES ('<(1,2),100>');
INSERT INTO CIRCLE_TBL VALUES ('1,3,5');
INSERT INTO CIRCLE_TBL VALUES ('((1,2),3)');
INSERT INTO CIRCLE_TBL VALUES ('<(100,200),10>');
INSERT INTO CIRCLE_TBL VALUES ('<(100,1),115>');
-- bad values
INSERT INTO CIRCLE_TBL VALUES ('<(-100,0),-100>');
ERROR:  invalid input syntax for type circle: "<(-100,0),-100>"
LINE 1: INSERT INTO CIRCLE_TBL VALUES ('<(-100,0),-100>');
                                       ^
INSERT INTO CIRCLE_TBL VALUES ('1abc,3,5');
ERROR:  invalid input syntax for type circle: "1abc,3,5"
LINE 1: INSERT INTO CIRCLE_TBL VALUES ('1abc,3,5');
                                       ^
INSERT INTO CIRCLE_TBL VALUES ('(3,(1,2),3)');
ERROR:  invalid input syntax for type circle: "(3,(1,2),3)"
LINE 1: INSERT INTO CIRCLE_TBL VALUES ('(3,(1,2),3)');
                                       ^
SELECT * FROM CIRCLE_TBL;
       f1       
----------------
 <(5,1),3>
 <(1,2),100>
 <(1,3),5>
 <(1,2),3>
 <(100,200),10>
 <(100,1),115>
(6 rows)

SELECT '' AS six, center(f1) AS center
  FROM CIRCLE_TBL;
 six |  center   
-----+-----------
     | (5,1)
     | (1,2)
     | (1,3)
     | (1,2)
     | (100,200)
     | (100,1)
(6 rows)

SELECT '' AS six, radius(f1) AS radius
  FROM CIRCLE_TBL;
 six | radius 
-----+--------
     |      3
     |    100
     |      5
     |      3
     |     10
     |    115
(6 rows)

SELECT '' AS six, diameter(f1) AS diameter
  FROM CIRCLE_TBL;
 six | diameter 
-----+----------
     |        6
     |      200
     |       10
     |        6
     |       20
     |      230
(6 rows)

SELECT '' AS two, f1 FROM CIRCLE_TBL WHERE radius(f1) < 5;
 two |    f1     
-----+-----------
     | <(5,1),3>
     | <(1,2),3>
(2 rows)

SELECT '' AS four, f1 FROM CIRCLE_TBL WHERE diameter(f1) >= 10;
 four |       f1       
------+----------------
      | <(1,2),100>
      | <(1,3),5>
      | <(100,200),10>
      | <(100,1),115>
(4 rows)

SELECT '' as five, c1.f1 AS one, c2.f1 AS two, (c1.f1 <-> c2.f1) AS distance
  FROM CIRCLE_TBL c1, CIRCLE_TBL c2
  WHERE (c1.f1 < c2.f1) AND ((c1.f1 <-> c2.f1) > 0)
  ORDER BY distance, area(c1.f1), area(c2.f1);
 five |      one       |      two       |     distance     
------+----------------+----------------+------------------
      | <(100,200),10> | <(100,1),115>  |               74
      | <(100,200),10> | <(1,2),100>    | 111.370729772479
      | <(1,3),5>      | <(100,200),10> | 205.476756144497
      | <(5,1),3>      | <(100,200),10> |  207.51303816328
      | <(1,2),3>      | <(100,200),10> | 208.370729772479
(5 rows)

--
-- Test the SP-GiST index
--
CREATE TEMPORARY TABLE quad_circle_tbl (id int, c circle);
INSERT INTO quad_circle_tbl
	SELECT (x - 1) * 100 + y, circle(point(x * 10, y * 10), 1 + (x + y) % 10)
	FROM generate_series(1, 100) x,
		 generate_series(1, 100) y;
INSERT INTO quad_circle_tbl
	SELECT i, '<(200, 300), 5>'
	FROM generate_series(10001, 11000) AS i;
INSERT INTO quad_circle_tbl
	VALUES
		(11001, NULL),
		(11002, NULL),
		(11003, '<(0,100), infinity>'),
		(11004, '<(-infinity,0),1000>'),
		(11005, '<(infinity,-infinity),infinity>');
CREATE INDEX quad_circle_tbl_idx ON quad_circle_tbl USING spgist(c);
-- get reference results for ORDER BY distance from seq scan
SET enable_seqscan = ON;
SET enable_indexscan = OFF;
SET enable_bitmapscan = OFF;
CREATE TEMP TABLE quad_circle_tbl_ord_seq1 AS
SELECT rank() OVER (ORDER BY c <-> point '123,456') n, c <-> point '123,456' dist, id
FROM quad_circle_tbl;
CREATE TEMP TABLE quad_circle_tbl_ord_seq2 AS
SELECT rank() OVER (ORDER BY c <-> point '123,456') n, c <-> point '123,456' dist, id
FROM quad_circle_tbl WHERE c <@ circle '<(300,400),200>';
-- check results results from index scan
SET enable_seqscan = OFF;
SET enable_indexscan = OFF;
SET enable_bitmapscan = ON;
EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c << circle '<(300,400),200>';
                         QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c << '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c << '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c << circle '<(300,400),200>';
 count 
-------
   891
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c &< circle '<(300,400),200>';
                         QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c &< '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c &< '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c &< circle '<(300,400),200>';
 count 
-------
  5901
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c && circle '<(300,400),200>';
                         QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c && '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c && '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c && circle '<(300,400),200>';
 count 
-------
  2334
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c &> circle '<(300,400),200>';
                         QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c &> '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c &> '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c &> circle '<(300,400),200>';
 count 
-------
 10000
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c >> circle '<(300,400),200>';
                         QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c >> '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c >> '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c >> circle '<(300,400),200>';
 count 
-------
  4990
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c <<| circle '<(300,400),200>';
                         QUERY PLAN                          
-------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c <<| '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c <<| '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c <<| circle '<(300,400),200>';
 count 
-------
  1890
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c &<| circle '<(300,400),200>';
                         QUERY PLAN                          
-------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c &<| '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c &<| '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c &<| circle '<(300,400),200>';
 count 
-------
  6900
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c |&> circle '<(300,400),200>';
                         QUERY PLAN                          
-------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c |&> '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c |&> '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c |&> circle '<(300,400),200>';
 count 
-------
  9000
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c |>> circle '<(300,400),200>';
                         QUERY PLAN                          
-------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c |>> '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c |>> '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c |>> circle '<(300,400),200>';
 count 
-------
  3990
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c @> circle '<(300,400),1>';
                        QUERY PLAN                        
----------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c @> '<(300,400),1>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c @> '<(300,400),1>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c @> circle '<(300,400),1>';
 count 
-------
     2
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c <@ circle '<(300,400),200>';
                         QUERY PLAN                         
------------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c <@ '<(300,400),200>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c <@ '<(300,400),200>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c <@ circle '<(300,400),200>';
 count 
-------
  2181
(1 row)

EXPLAIN (COSTS OFF)
SELECT count(*) FROM quad_circle_tbl WHERE c ~= circle '<(300,400),1>';
                        QUERY PLAN                        
----------------------------------------------------------
 Aggregate
   ->  Bitmap Heap Scan on quad_circle_tbl
         Recheck Cond: (c ~= '<(300,400),1>'::circle)
         ->  Bitmap Index Scan on quad_circle_tbl_idx
               Index Cond: (c ~= '<(300,400),1>'::circle)
(5 rows)

SELECT count(*) FROM quad_circle_tbl WHERE c ~= circle '<(300,400),1>';
 count 
-------
     1
(1 row)

RESET enable_seqscan;
RESET enable_indexscan;
RESET enable_bitmapscan;
